judge-d-server/src/main/java/com/hltech/judged/server/infrastructure/persistence/environment/JPAEnvironmentRepository.java

```diff
@@ -0,0 +1,76 @@
+package com.hltech.judged.server.infrastructure.persistence.environment;
+
+import com.hltech.judged.server.domain.environment.Environment;
+import com.hltech.judged.server.domain.environment.EnvironmentRepository;
+import com.hltech.judged.server.domain.environment.Service;
+import com.hltech.judged.server.domain.environment.Space;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Repository;
+
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+@Repository
+@RequiredArgsConstructor
+public class JPAEnvironmentRepository implements EnvironmentRepository {
+
+    private final SpringDataEnvironmentRepository springDataEnvironmentRepository;
+
+    @Override
+    public Environment persist(Environment environment) {
+        EnvironmentTuple environmentTuple = toEnvironmentTuple(environment);
+        return toEnvironment(springDataEnvironmentRepository.saveAndFlush(environmentTuple));
+    }
+
+    @Override
+    public Set<String> getNames() {
+        return springDataEnvironmentRepository.getNames();
+    }
+
+    @Override
+    public Environment get(String name) {
+        return springDataEnvironmentRepository.findById(name)
+            .map(this::toEnvironment)
+            .orElse(new Environment(name, new HashSet<>()));
+    }
+
+    private EnvironmentTuple toEnvironmentTuple(Environment environment) {
+        return new EnvironmentTuple(environment.getName(), getSpaceServiceVersions(environment));
+    }
+
+    private Set<ServiceVersion> getSpaceServiceVersions(Environment environment) {
+        return environment.getSpaceNames().stream()
+            .flatMap(space -> environment.getServices(space).stream()
+                .map(service -> new ServiceVersion(space, service.getName(), service.getVersion())))
+            .collect(Collectors.toUnmodifiableSet());
+    }
+
+    private Environment toEnvironment(EnvironmentTuple environmentTuple) {
+        Set<Space> spaces = new HashSet<>();
+
+        environmentTuple.getServiceVersions()
```

Comment:
a lot of nested streams here, i would try to simplify this or at least extract some parts to separate methods